libkodimote/mpris2/mprisplayer.cpp

```diff
@@ -0,0 +1,394 @@
+#include "mprisplayer.h"
+
+#include "kodi.h"
+#include "libraryitem.h"
+#include "player.h"
+#include "playlist.h"
+
+MprisPlayer::MprisPlayer(ProtocolManager *protocols, QObject *parent) :
+    QDBusAbstractAdaptor(parent),
+    m_player(Kodi::instance()->activePlayer()),
+    m_protocols(protocols)
+{
+    setAutoRelaySignals(false);
+    connect(Kodi::instance(), SIGNAL(activePlayerChanged()), SLOT(activePlayerChanged()));
+    connect(Kodi::instance(), SIGNAL(volumeChanged(int)), SLOT(volumeChanged()));
+}
+
+bool MprisPlayer::canControl() const
+{
+    return true;
+}
+
+bool MprisPlayer::canGoNext() const
+{
+    if (!m_player) {
+        return false;
+    }
+
+    return m_player->playlist()->currentTrackNumber() < m_player->playlist()->count();
+}
+
+bool MprisPlayer::canGoPrevious() const
+{
+    if (!m_player) {
+        return false;
+    }
+
+    return m_player->playlist()->currentTrackNumber() > 1;
+}
+
+bool MprisPlayer::canPause() const
+{
+    return true;
+}
+
+bool MprisPlayer::canPlay() const
+{
+    if (!m_player) {
+        return false;
+    }
+
+    return !!m_player->currentItem();
+}
+
+bool MprisPlayer::canSeek() const
+{
+    return true;
+}
+
+double MprisPlayer::maximumRate() const
+{
+    return 32.0;
+}
+
+double MprisPlayer::minimumRate() const
+{
+    return -32.0;
+}
+
+QVariantMap MprisPlayer::metadata() const
+{
+    QVariantMap map;
+    if (m_player && m_player->currentItem()) {
+        LibraryItem *item = m_player->currentItem();
+        map["mpris:trackid"] = qVariantFromValue(QDBusObjectPath(buildPath(item)));
```

Comment:
This breaks the metadata call and is the reason why metadata doesn't show up in ubuntu. Removing the qVariantFromValue(QDBusObjectPath()) seems to fix it.